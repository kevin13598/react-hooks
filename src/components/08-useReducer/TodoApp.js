import React, { useReducer }  from 'react';
import { todoReducer } from './todoReducer';
import { useEffect } from 'react';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

import './styles.css';

const init = () => {

    return JSON.parse(localStorage.getItem('todos')) || [];

}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    

    useEffect( () => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = (id) => {
        
        dispatch({
            type: 'delete',
            payload: id
        });

    }

    const handleToggle = (id) => {
        dispatch({
            type: 'toggle',
            payload: id
        })
    }

    const handleAddTodo = (newTodo) => {

        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr/>

           <div className="row">

               <div className="col-7">

                    <TodoList 
                        todos={todos} 
                        onDelete={handleDelete} 
                        onToggle={handleToggle} 
                    />

               </div>

               <div className="col-5">
                    <TodoAdd handleAddtodo = {handleAddTodo} />
               </div>

           </div>

        </div>
    )
}
