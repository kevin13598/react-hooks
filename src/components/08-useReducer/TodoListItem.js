import React from 'react';

export const TodoListItem = ({todo, index, onToggle, onDelete}) => {

    return (
        <li
            key={todo.id}
            className="list-group-item"
        >
            <p 
                className={`${todo.done && 'complete'}`}
                onClick= {() => onToggle(todo.id)}
            >
                {index+1}. {todo.desc}
            </p>

            <button
                className="btn btn-danger"
                onClick = {() => onDelete(todo.id)}
            >
                Borrar
            </button>
    </li>                     
    )
}
