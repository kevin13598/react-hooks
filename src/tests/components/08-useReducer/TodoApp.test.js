import {TodoApp} from "../../../components/08-useReducer/TodoApp";
import { shallow, mount } from "enzyme";
import React from 'react';
import { demoTodos } from "../../fixtures/demoTodos";
import { act } from "@testing-library/react";

describe('Pruebas en <TodoApp/>', () => {

    const wrapper = shallow(<TodoApp/>);

    Storage.prototype.setItem = jest.fn(() => {});

    test('debe de mostrarse el componente correctamente', () => {
        
        expect(wrapper).toMatchSnapshot();

    });

    test('debe de agregar un TODO', () => {
        
        const wrapper = mount(<TodoApp />);

        act( () => {
            wrapper.find('TodoAdd').prop('handleAddtodo')(demoTodos[0]);
            wrapper.find('TodoAdd').prop('handleAddtodo')(demoTodos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);

    });

    test('debe de eliminar un todo', () => {
        
        act( () => {
            wrapper.find('TodoAdd').prop('handleAddtodo')(demoTodos[0]);
            wrapper.find('TodoList').prop('onDelete')(demoTodos[0].id);
        });
        

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)');


    });
    

});