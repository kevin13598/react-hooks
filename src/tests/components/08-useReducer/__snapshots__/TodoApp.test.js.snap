// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoApp/> debe de mostrarse el componente correctamente 1`] = `
<div>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        onDelete={[Function]}
        onToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddtodo={[Function]}
      />
    </div>
  </div>
</div>
`;
